src/replicache.ts

```diff
@@ -1237,6 +1258,201 @@ export class Replicache<MD extends MutatorDefs = {}> {
       return {result, ref};
     });
   }
+
+  private async _recoverMutations(): Promise<void> {
+    if (this._recoveringMutations) {
+      return;
+    }
+    if (!this.online) {
+      return;
+    }
+    try {
+      this._recoveringMutations = true;
+      const ok = await this._recoverMutationsFromPerdag(
+        this._perdag,
+        this.schemaVersion,
+      );
+      if (ok) {
+        for (const schemaVersion of this.previousSchemaVersions) {
+          const perKvStore = new IDBStore(
+            makeIdbName(this.name, this.schemaVersion),
+          );
+          const perdag = new dag.StoreImpl(
+            perKvStore,
+            dag.throwChunkHasher,
+            assertNotTempHash,
+          );
+          const ok = await this._recoverMutationsFromPerdag(
+            perdag,
+            schemaVersion,
+          );
+          if (!ok) {
+            break;
```

Comment:
Not sure about this. Why should it exit the loop here?

Reply:
The idea is to fail fast on the mutation recovery process if we hit a connectivity error.  _recoverMutationsFromPerdag returns false only when hitting a connectivity error.  When we transition from offline -> online we will restart the process by calling this._recoverMutations.

I consider it an error if push or pull got an error and this.online is now false.  For other errors like the server returning a 500, we do not fail fast.

It may be more reliable to not fail fast.  If for some reason recovering a particular client consistently causes an error that we incorrectly interpret as a connectivity error, we will never recover clients after it in the ClientMap.  The tradeoff is we do some perdag reading and pushes that we expect to fail because we are "offline".

Reply:
decided to simplify and not fail fast on connectivity issues, this is simpler and more robust.